Improve performance in the WP_Theme_JSON class by simplifying calls to array functions

aristath opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/56974

This PR simplifies expensive array-functions calls where appropriate.

Attaching some profiling results to showcase the impact.

Before (trunk):
<img width="950" alt="Screenshot 2022-11-03 at 9 43 25 AM" src="https://user-images.githubusercontent.com/588688/199669150-1fcf3ccc-f0fc-49b7-b54d-9a4eb3635398.png">

After (with this PR):
<img width="950" alt="Screenshot 2022-11-03 at 9 43 15 AM" src="https://user-images.githubusercontent.com/588688/199669142-886ff31f-ab79-43fb-801a-94a9d61892d5.png">

All numbers in the screenshots above use milliseconds, and the impact of these seemingly simple changes is bigger than expected, shaving off more than 200ms. Of course those numbers are a bit inflated due to the fact that profiling slows down a page significantly, but the overall improvement on the page-load is more than 3%.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

props @jrfnl 🙌

jrfnl committed:
WIP

jrfnl committed:
WIP step 2

jrfnl committed:
WIP - step 3

jrfnl committed:
WIP Step 3b

jrfnl committed:
WIP - setp 4

jrfnl committed:
WIP - simplify it a little more

jrfnl committed:
WIP - step 5

jrfnl committed:
WIP - side step 6

jrfnl committed:
WIP - side step 7

aristath committed:
CS & cleanup

aristath committed:
replace array_key_exists with isset

These 2 functions only differ if the value in one of the items of the array has a value of null - something that can't ever happen in this scenario. isset() is significantly faster so changing it makes sense in this context.

aristath committed:
replace an array_merge with a foreach loop

aristath committed:
Add an inline comment to avoid future refactors

aristath committed:
We can use isset() instead of array_key_exists() here (no null values)

aristath committed:
replace array_merge with a foreach loop

aristath committed:
replace array_merge with foreach loop in the flatten_tree method

aristath committed:
replace array_key_exists() with isset()

Checked each of these cases to ensure that value is never null

aristath committed:
Replace more array_merge with foreach loops where appropriate

jrfnl commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -456,7 +456,7 @@ class WP_Theme_JSON {
 		$class_name = '';
 
-		if ( array_key_exists( $element, static::__EXPERIMENTAL_ELEMENT_CLASS_NAMES ) ) {
+		if ( isset( static::__EXPERIMENTAL_ELEMENT_CLASS_NAMES[ $element ] ) ) {
```
> Fatal error: Cannot use isset() on the result of an expression (you can use "null !== expression" instead) in ./src/wp-includes/class-wp-theme-json.php on line 459

jrfnl replied:
☝🏻 This is your PHP 5.6 error ;-)

spacedmonkey replied:
https://github.com/WordPress/gutenberg/pull/42567

jrfnl replied:
Just as a note for the record: running `isset()`on plain constants/class constants is fine, it's only array access on those constants which is problematic.

The ability to declare a `const` as an array was only added in PHP 5.6, the ability to do the same using `define()` was added in PHP 7.0, which is probably why the ability to use `isset()` on array access on constants was added in PHP 7.0.

aristath replied:
Perfect, thank you! I reverted the lines with the 5.6 issue, and added notes so we know what to do in the future after WP drops support for PHP 5.6 😄

aristath committed:
Fix PHP 5.6 errors + added notes for the future

spacedmonkey commented:
While doing some profiling, I noticed there are still calls to `_wp_array_get`. This calls `array_key_exists`, which is expensive. I wonder if we could avoid these somehow.

spacedmonkey commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2317,6 +2343,9 @@ public function merge( $incoming ) {
 						_wp_array_set( $this->theme_json, $path, $content );
 					} else {
+						$slugs_node = static::get_default_slugs( $this->theme_json, $node['path'] );
+						$slugs      = array_merge_recursive( $slugs_global, $slugs_node );
```
Why is this needed?

aristath replied:
It wasn't _added_, it was just _moved_ here. 😉 
A few lines above we were calling  these 2 on the previous `foreach` loop (used to be line 2281), so they were always running. Instead of always calling them, we moved them here where they are actually used. This way these rather expensive calls run fewer times, only when necessary.

jrfnl commented:
> While doing some profiling, I noticed there are still calls to `_wp_array_get`. This calls `array_key_exists`, which is expensive. I wonder if we could avoid these somehow.

I wouldn't worry too much about (or overestimate) the impact of `array_key_exists()`.
* `array_key_exists()` is **_not_** one of the typically expensive `array_*()` function calls.
* Yes, `isset()` is generally speaking faster...
* ... BUT `array_key_exists()` has been [opcode optimized in PHP since PHP 7.4](https://speakerdeck.com/jrf/for-the-love-of-code-modernizing-wordpress-plugins-and-themes-full-deck?slide=45) and what with the [fast majority of WP users being on PHP 7.4 (or above)](https://wordpress.org/about/stats/), the performance difference will be negligible for most users.

Note: `isset()` and `array_key_exists()` also have a _functional_ difference in how they treat a `null` value, so for `array_key_exists()` to be equivalent to `isset()`, it MUST be accompanied by a `&& $var !== null`.

aristath commented:
> Note: `isset()` and `array_key_exists()` also have a _functional_ difference in how they treat a `null` value, so for `array_key_exists()` to be equivalent to `isset()`, it MUST be accompanied by a `&& $var !== null`.

Yeah, that's an important distinction!
In the comments I added for future reference to replace `array_key_exists()` with `isset()` (`// Note: array_key_exists() should be replaced with an isset() check once WordPress drops support for PHP 5.6.`), I checked the values in each case and only added that comment in cases where the value can not be `null`. 
In one case the value _can_ be `null`, and I added a comment in that one that `// We need to use array_key_exists() instead of isset() because the value can be null.` 👍

spacedmonkey commented:
In trunk. 
<img width="503" alt="Screenshot 2022-11-09 at 21 54 06" src="https://user-images.githubusercontent.com/237508/200950234-4d0c7513-3721-4640-ac26-a74b6cf02d69.png">

In this brunch. 
<img width="525" alt="Screenshot 2022-11-09 at 21 54 15" src="https://user-images.githubusercontent.com/237508/200950270-30fd48a7-ac96-4cb4-a541-ab28e5684e78.png">

felixarntz commented:
@jrfnl @aristath Following up on `array_key_exists`: Makes sense not to overthink this in terms of performance impact compared to `isset`, however I had one follow up thought I wanted to at least raise here:
* Most of the usage of `array_key_exists` here could be replaced with `isset` as the additional `null` consideration is irrelevant for most (if not all) of the code where it's used here.
* Given that, there would be the alternative option of assigning the result to a variable and then calling `isset` on that variable.

For example, instead of:
```php
if ( array_key_exists( $element, static::__EXPERIMENTAL_ELEMENT_CLASS_NAMES ) ) {
}
```

Do:
```php
$element_class_names = static::__EXPERIMENTAL_ELEMENT_CLASS_NAMES;
if ( isset( $element_class_names[ $element ] ) {
}
```

Again, not saying this is necessarily worth doing (and certainly not a blocker), but I'd be curious if the first or the last is generally faster. :)

Commit:
Merge branch 'trunk' into feature/pp-ari-wp-theme-json

felixarntz reviewed (approved):
The code changes here look solid to me!

I of course trust the benchmarking data provided in https://core.trac.wordpress.org/ticket/56974#comment:1, but it may additionally be helpful to quantify the performance gain here in milliseconds as well.

Given that parsing `theme.json` affects the `wp_head` action quite a bit (see also #3536), I thought I'd do a performance comparison there first. Based on 9 test runs, the median impact there is small but consistent:
* With Twenty Twenty-Three (i.e. block theme): 26.63ms vs 27.78ms (~4% faster)
* With Twenty Twenty-One (i.e. classic theme): 15.30ms vs 16.05ms (~5% faster)

See [this spreadsheet](https://docs.google.com/spreadsheets/d/1PSe9wx3xSnyQNS5w_NJUrHXo-txweOvt_RXCeYrvcLU/edit) for the full data.

So also in that regard it looks like a small but worthwhile win. Potentially I can do the same for more concretely the parsing of `theme.json` itself as that would be even closer to the changes here and allow us to quantify impact in that way too.

jrfnl commented:
There's bound to be more room for improvement, but yes, let's take this step by step and I do believe this is a valuable step in that.

Re: assign+`isset` vs `array_key_exists`: as most `array_key_exists()` are only in place for PHP 5.6, might be more effective to re-open the discussion on dropping support for PHP 5.6 ... 😇 

On that note - PHP 8.3 will bring another very useful improvement [`json_validate()`](https://wiki.php.net/rfc/json_validate) (validate a JSON input before decoding without running the risk of the JSON tying up all the memory), but that is no doubt still a bridge too far ;-)

desrosj commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -603,6 +607,7 @@ protected static function sanitize( $input, $valid_block_names, $valid_element_n
 		 */
 		$styles_non_top_level = static::VALID_STYLES;
 		foreach ( array_keys( $styles_non_top_level ) as $section ) {
+			// We need to use array_key_exists() instead of isset() because the value can be null.
```
Could this be reworded to remove "we"?

felixarntz replied:
Updated.

desrosj commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -456,6 +456,7 @@ class WP_Theme_JSON {
 	public static function get_element_class_name( $element ) {
 		$class_name = '';
 
+		// Note: array_key_exists() should be replaced with an isset() check once WordPress drops support for PHP 5.6.
```
Could a Trac ticket be opened for tracking this with a detailed explanation of why that can be linked to for additional context? At a minimum, a summary of the discussion on this PR into a comment on the current Trac ticket would also work.

felixarntz replied:
Opened https://core.trac.wordpress.org/ticket/57067 and updated all of these comments throughout the PR to point to it.

desrosj commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -626,6 +631,7 @@ protected static function sanitize( $input, $valid_block_names, $valid_element_n
 		foreach ( $valid_element_names as $element ) {
 			$schema_styles_elements[ $element ] = $styles_non_top_level;
 
+			// Note: array_key_exists() should be replaced with an isset() check once WordPress drops support for PHP 5.6.
```
Same here for using `See: URL.` to link off for additional details/context.

There are several more throughout the changes here, but I'm going to stop commenting on each one inline.

desrosj commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -1662,6 +1672,7 @@ protected static function compute_style_properties( $styles, $settings = array()
 			if ( is_array( $value_path ) ) {
 				$path_string = implode( '.', $value_path );
 				if (
+					// Note: array_key_exists() should be replaced with an isset() check once WordPress drops support for PHP 5.6.
```
Same here for using `See: URL.` to link off for additional details/context.

desrosj reviewed (changes requested):
Added a few nits.

Commit:
Add comment references pointing to https://core.trac.wordpress.org/ticket/57067.

Commit:
Merge branch 'trunk' into feature/pp-ari-wp-theme-json

Commit:
Merge branch 'trunk' into feature/pp-ari-wp-theme-json

desrosj commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -1273,8 +1281,12 @@ protected function get_css_variables( $nodes, $origins ) {
+			$node                    = _wp_array_get( $this->theme_json, $metadata['path'], array() );
+			$declarations            = static::compute_preset_vars( $node, $origins );
+			$theme_vars_declarations = static::compute_theme_vars( $node );
+			foreach ( $theme_vars_declarations as $declaration ) {
```
I got a bit tripped up on these variable names. `$declaration` feels like `$declarations` is being iterated, but that's not the case here. Maybe `$theme_var_declaration` instead? Don't really feel strongly about this, it just read a bit weird to me.

felixarntz replied:
SG, will update.

desrosj commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -519,7 +521,10 @@ public function __construct( $theme_json = array(), $origin = 'theme' ) {
-				$path   = array_merge( $node['path'], $preset_metadata['path'] );
+				$path = $node['path'];
+				foreach ( $preset_metadata['path'] as $subpath ) {
+					$path[] = $subpath;
```
Just noting that this doesn't protect against duplicate values being added to `$path`, but that's how it works currently as `array_merge()` also results in duplicates values being present.

felixarntz replied:
Right, was looking at this too, but indeed not a change in behavior, so this is fine as is.

desrosj commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2283,11 +2310,11 @@ public function merge( $incoming ) {
-
 			// Replace the spacing.units.
-			$path    = array_merge( $node['path'], array( 'spacing', 'units' ) );
+			$path   = $node['path'];
```
Why not just use an array declaration here instead of three individual assignments?

```
$path = array(
	$node['path'],
	'spacing',
	'units'
);
```

felixarntz replied:
Your suggested code would not be the same result. `$node['path']` is an array, so in your code the first element would be an array and the other two would be strings, so that would not be correct I think.

desrosj replied:
Ah, yes. You're right on this one. I think a few others it may result in the same. But not this one.

desrosj commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2298,19 +2325,25 @@ public function merge( $incoming ) {
+						$base_path[] = $leaf;
+					}
+
+					$path   = $base_path;
```
Same question here about why not use `$path = array()`;

felixarntz replied:
Based on what the original code does, `$path` needs to include the items in `$base_path`, plus the single `$origin` item, so if we set `$path` to an empty array here, the behavior would change.

desrosj commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2463,7 +2504,8 @@ protected static function get_default_slugs( $data, $node_path ) {
 	 */
 	protected function get_name_from_defaults( $slug, $base_path ) {
-		$path            = array_merge( $base_path, array( 'default' ) );
+		$path            = $base_path;
```
Same question about `$path = array();`

felixarntz replied:
See above, that would change the behavior.

desrosj commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2593,8 +2637,9 @@ protected static function remove_insecure_settings( $input ) {
 			foreach ( static::VALID_ORIGINS as $origin ) {
-				$path_with_origin = array_merge( $preset_metadata['path'], array( $origin ) );
-				$presets          = _wp_array_get( $input, $path_with_origin, null );
+				$path_with_origin   = $preset_metadata['path'];
```
Can we use `$path_with_origin = array( value1, value2 )` here as well?

felixarntz replied:
Similar to https://github.com/WordPress/wordpress-develop/pull/3555#discussion_r1019751990, that would change the data, since `$preset_metadata['path']` is an array itself, so you would end up with an array in an array.

desrosj reviewed (approved):
Looks good overall. Commented with a few small questions/suggestions, but none of them are really to be considered blockers.

Commit:
Merge branch 'trunk' into feature/pp-ari-wp-theme-json

Commit:
Clarify variable name.

felixarntz commented:
Committed in https://core.trac.wordpress.org/changeset/54804.

oandregal commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -2035,7 +2056,9 @@ public function get_styles_for_block( $block_metadata ) {
 					if ( isset( $feature_declarations[ $feature_selector ] ) ) {
-						$feature_declarations[ $feature_selector ] = array_merge( $feature_declarations[ $feature_selector ], $new_feature_declarations );
+						foreach ( $new_feature_declarations as $new_feature_declaration ) {
+							$feature_declarations[ $feature_selector ][] = $feature_declaration;
```
@aristath @desrosj @felixarntz @spacedmonkey @jrfnl @hellofromtonya 

Shouldn't the variable assigned be `$new_feature_declaration` instead of `$feature_declaration`? This has been reported as problematic by the Gutenberg lint jobs ([see conversation](https://github.com/WordPress/gutenberg/pull/46579/files#r1049600844)).

I haven't looked at what regression this introduced but it'd be good to create a follow-up PR fixing it and adding a test case.

jrfnl replied:
@oandregal I concur. This is a bug which should be fixed.
```suggestion
							$feature_declarations[ $feature_selector ][] = $new_feature_declaration;
```

aristath replied:
🤦 yeah, looks like this was a typo.. Good catch 👍

jrfnl replied:
Happens to the best of us ;-)

oandregal replied:
I'm not familiar with this code path, so asked folks to provide details for testing at https://github.com/WordPress/gutenberg/pull/46579#discussion_r1049600844 so we can add a new unit test and gauge how severe this is.

jrfnl replied:
AFAICS, this should go in the next 6.1 patch release.

Edit to clarify: I'm basing this assessment on the following:
* It's very clearly a bug with an obvious fix.
* It's likely a regression compared to previous behaviour.
